### PR TITLE
generating and filling a random state test

### PR DESCRIPTION
**Pull request by winsvega**
Usage: 

./testeth -t StateTestsGeneral -- --createRandomTest

### CODE REVIEW

File: test/tools/fuzzTesting/createRandomTest.cpp [hunk 1]
```diff
@@ -50,62 +50,64 @@ int createRandomTest(std::vector<char*> const& _parameters)
...
+		if (arg == "--debug")
 			debug = true;
 		else
 		if (arg == "--filldebug")
 			filldebug = true;
 	}
 
-	if (testSuite == "")
+	if (checktest)
+		std::cout << "Testing: " << testSuite.substr(0, testSuite.length() - 1) << "... ";
```
Comment (pirapira) [inline test/tools/fuzzTesting/createRandomTest.cpp]: What does this `substr` do?

File: test/tools/libtesteth/TestOutputHelper.cpp [hunk 2]
```diff
@@ -32,7 +32,7 @@ using namespace boost;
 
 Timer TestOutputHelper::m_timer;
 size_t TestOutputHelper::m_currTest = 0;
-size_t TestOutputHelper::m_maxTests = 0;
+size_t TestOutputHelper::m_maxTests = 1;
```
Comment (pirapira) [inline test/tools/libtesteth/TestOutputHelper.cpp]: What does this change do?

File: test/tools/libtesteth/TestOutputHelper.cpp [hunk 3]
```diff
@@ -44,7 +44,8 @@ void TestOutputHelper::initTest(int _maxTests)
 	NoProof::init();
 	m_timer.restart();
 	m_currentTestCaseName = boost::unit_test::framework::current_test_case().p_name;
-	std::cout << "Test Case \"" + m_currentTestCaseName + "\": " << std::endl;
+	if (!Options::get().createRandomTest)
+		std::cout << "Test Case \"" + m_currentTestCaseName + "\": " << std::endl;
```
Comment (pirapira) [inline test/tools/libtesteth/TestOutputHelper.cpp]: Should it output a different name when `createRandomTest`?

File: test/tools/fuzzTesting/createRandomTest.cpp [hunk 4]
```diff
@@ -50,62 +50,64 @@ int createRandomTest(std::vector<char*> const& _parameters)
...
 	{
-		std::cerr << "Error! Test suite not supported! (Usage -t TestSuite)" << std::endl;
-		return 1;
+		if (checktest)
+			return checkRandomTest(dev::test::doStateTests, testmValue, debug);
+		else
+			fillRandomTest(dev::test::doStateTests, (filltest) ? testFillString : c_testExampleStateTest, filldebug);
 	}
 	else
+	if (testSuite == "BlockchainTests")
```
Comment (pirapira) [inline test/tools/fuzzTesting/createRandomTest.cpp]: Maybe `else if` should be on the same line.

File: test/tools/fuzzTesting/createRandomTest.cpp [hunk 5]
```diff
@@ -50,62 +50,64 @@ int createRandomTest(std::vector<char*> const& _parameters)
...
-	if (testSuite == "")
+	if (checktest)
+		std::cout << "Testing: " << testSuite.substr(0, testSuite.length() - 1) << "... ";
+
+	if (testSuite == "StateTestsGeneral")
 	{
-		std::cerr << "Error! Test suite not supported! (Usage -t TestSuite)" << std::endl;
-		return 1;
+		if (checktest)
+			return checkRandomTest(dev::test::doStateTests, testmValue, debug);
```
Comment (pirapira) [inline test/tools/fuzzTesting/createRandomTest.cpp]: Why is `fillRandomTest` is skipped when `chekcktest`?   Isn't `checktest` similar to `checkstate` option?  I felt the filling should happen even with the `checktest` option.

Also, similarly to `checkstate` option, maybe `checktest` should always be true.

File: test/tools/libtesteth/TestOutputHelper.cpp (hunk 2 above)
Comment (winsvega) [inline test/tools/libtesteth/TestOutputHelper.cpp]: there was a division by 0 if TestOutputHelper not being initialized. with init function

File: test/tools/libtesteth/TestOutputHelper.cpp (hunk 3 above)
Comment (winsvega) [inline test/tools/libtesteth/TestOutputHelper.cpp]: createRandomTest provide a clean json object to the std output

File: test/tools/fuzzTesting/createRandomTest.cpp (hunk 5 above)
Comment (winsvega) [inline test/tools/fuzzTesting/createRandomTest.cpp]: this is a legacy code and actually is not working. 
there was an option to run the test provided by std input. 

this PR enables the support of 
```./testeth -t StateTestsGeneral -- --createRandomTest```
command 
other options like --checkstate are yet not supported. but I could review the code when needed.

Comment (pirapira) [inline test/tools/fuzzTesting/createRandomTest.cpp]: I see.  In that case, let's remove the legacy code before adding new features.

File: test/tools/libtesteth/TestOutputHelper.cpp (hunk 3 above)
Comment (pirapira) [inline test/tools/libtesteth/TestOutputHelper.cpp]: OK.  I didn't know this.

File: test/tools/fuzzTesting/createRandomTest.cpp (hunk 5 above)
Comment (winsvega) [inline test/tools/fuzzTesting/createRandomTest.cpp]: this new features are actually revived old ones.

File: test/tools/fuzzTesting/createRandomTest.cpp [hunk 6]
```diff
@@ -50,62 +50,64 @@ int createRandomTest(std::vector<char*> const& _parameters)
...
+		else
+			fillRandomTest(dev::test::doStateTests, (filltest) ? testFillString : c_testExampleStateTest, filldebug);
```
Comment (gumb0) [inline test/tools/fuzzTesting/createRandomTest.cpp]: `filltest` is always false, remove it alltogether, it will simplify things

File: test/tools/fuzzTesting/createRandomTest.cpp [hunk 7]
```diff
@@ -218,42 +204,69 @@ void dev::test::RandomCode::parseTestWithTypes(std::string& _test, std::map<std:
 				int randomDepth = 1 + (int)dev::test::RandomCode::randomUniInt() % 10;
 				_test.replace(pos, 5, dev::test::RandomCode::rndRLPSequence(randomDepth, debug));
 				cnote << debug;
+				std::string a;
```
Comment (gumb0) [inline test/tools/fuzzTesting/createRandomTest.cpp]: what's this?

File: test/tools/fuzzTesting/createRandomTest.cpp (hunk 5 above)
Comment (pirapira) [inline test/tools/fuzzTesting/createRandomTest.cpp]: Is there a way to make `checktest` true?

File: test/tools/fuzzTesting/createRandomTest.cpp [hunk 8]
```diff
@@ -262,7 +275,8 @@ void dev::test::RandomCode::parseTestWithTypes(std::string& _test, std::map<std:
 
 std::vector<std::string> dev::test::RandomCode::getTypes()
 {
-	return {"[RLP]", "[CODE]", "[HEX]", "[HEX32]", "[HASH20]", "[HASH32]", "[0xHASH32]", "[V]", "[GASLIMIT]"};
+	//declare possible types
+	return {"[RLP]", "[CODE]", "[HEX]", "[HEX32]", "[HASH20]", "[HASH32]", "[0xHASH32]", "[V]", "[GASLIMIT]", "[ADDRESS]", "[0xADDRESS]", "[0xDESTADDRESS]"};
```
Comment (gumb0) [inline test/tools/fuzzTesting/createRandomTest.cpp]: Make it a constant on top of the file instead of `getTypes()` function

File: test/tools/fuzzTesting/createRandomTest.cpp [hunk 9]
```diff
@@ -218,42 +204,69 @@ void dev::test::RandomCode::parseTestWithTypes(std::string& _test, std::map<std:
...
 				cnote << debug;
+				std::string a;
 			}
 			else
 			if (types.at(i) == "[CODE]")
-				_test.replace(pos, 6, "0x"+dev::test::RandomCode::generate(10, options));
+			{
+				int random = (int)dev::test::RandomCode::randomUniInt() % 100;
+				if (random < 90)
+					_test.replace(pos, types.at(i).length(), "0x"+dev::test::RandomCode::generate(10, options));
```
Comment (gumb0) [inline test/tools/fuzzTesting/createRandomTest.cpp]: Don't repeat `types.at(i)` 1000 times, use instead the loop `for (auto const& type: types)`

File: test/tools/fuzzTesting/createRandomTest.cpp [hunk 10]
```diff
@@ -42,71 +42,29 @@ int checkRandomTest(std::function<void(json_spirit::mValue&, bool)> _doTests, js
...
-			{
-				dev::eth::VMFactory::setKind(dev::eth::VMKind::JIT);
-				return checkRandomTest(dev::test::doVMTests, testmValue, debug);
-			}
-			else
-				fillRandomTest(dev::test::doVMTests, (filltest) ? testFillString : c_testExampleVMTest, filldebug);
-		}
-	}
+	if (testSuite == "StateTestsGeneral")
+		fillRandomTest(dev::test::doStateTests, c_testExampleStateTest, debug);
```
Comment (pirapira) [inline test/tools/fuzzTesting/createRandomTest.cpp]: Why doesn't VMTest appear here?  Maybe we should say "Test suite VMTest not supported" above?

File: test/tools/fuzzTesting/createRandomTest.cpp [hunk 11]
```diff
@@ -167,10 +125,13 @@ void fillRandomTest(std::function<void(json_spirit::mValue&, bool)> _doTests, st
 		dev::test::RandomCode::parseTestWithTypes(newTest, nullReplaceMap);
 		json_spirit::read_string(newTest, v);
 		_doTests(v, true);
+		for (auto& obj: v.get_obj())
+			if (!obj.second.get_obj().count("post"))
+				BOOST_ERROR("An error occured when executing random state test!");
```
Comment (pirapira) [inline test/tools/fuzzTesting/createRandomTest.cpp]: Can we make `_doTests` return a boolean indicating an error?

File: test/tools/fuzzTesting/createRandomTest.cpp [hunk 12]
```diff
@@ -185,84 +146,92 @@ void fillRandomTest(std::function<void(json_spirit::mValue&, bool)> _doTests, st
...
-		std::size_t pos = _test.find(types.at(i));
+		std::size_t pos = _test.find(type);
 		while (pos != std::string::npos)
 		{
-			if (types.at(i) == "[RLP]")
+			if (type == "[RLP]")
 			{
 				std::string debug;
 				int randomDepth = 1 + (int)dev::test::RandomCode::randomUniInt() % 10;
 				_test.replace(pos, 5, dev::test::RandomCode::rndRLPSequence(randomDepth, debug));
```
Comment (pirapira) [inline test/tools/fuzzTesting/createRandomTest.cpp]: Replace 5 with type.length().

File: test/tools/fuzzTesting/createRandomTest.cpp [hunk 13]
```diff
@@ -185,84 +146,92 @@ void fillRandomTest(std::function<void(json_spirit::mValue&, bool)> _doTests, st
...
+					_test.replace(pos, type.length(), "0x" + dev::test::RandomCode::rndByteSequence(1));
 			}
 			else
-				_test.replace(pos, types.at(i).size(), _varMap.at(types.at(i)));
+			{
+				//Replace type from varMap if varMap is set
+				if (_varMap.count(type))
+					_test.replace(pos, type.length(), _varMap.at(type));
+				BOOST_ERROR("Skipping undeclared type: " + type);
+			}
```
Comment (pirapira) [inline test/tools/fuzzTesting/createRandomTest.cpp]: You can siplify the code by defining `string replacement`.  In each case, you just do `replacement = something;`

After all these cases, a single
```
_test.replace(pos, type.length(), replacement);
```
will do the job.  This way, the many repetitions of `_test.replace(pos, type.length(), ....);` can be removed.

File: test/tools/fuzzTesting/fuzzHelper.cpp [hunk 14]
```diff
@@ -303,15 +303,15 @@ std::string RandomCode::generate(int _maxOpNumber, RandomCodeOptions _options)
 	return code;
 }
 
-std::string RandomCode::randomUniIntHex(u256 _maxVal)
+std::string RandomCode::randomUniIntHex(u256 _maxVal, u256 _minVal)
```
Comment (pirapira) [inline test/tools/fuzzTesting/fuzzHelper.cpp]: `_minVal` is never modified.  Consider `u256 const& _minVal`.

File: test/tools/libtesteth/Options.cpp [hunk 15]
```diff
@@ -99,6 +100,11 @@ Options::Options(int argc, char** argv)
 			if (i + 1 >= argc)
 				BOOST_THROW_EXCEPTION(InvalidOption(arg + " option is missing an argument."));
 		};
+		auto throwIfNotRandomTest = [&seenCreateRandomTest, &arg]()
+		{
+			if (!seenCreateRandomTest)
+				BOOST_THROW_EXCEPTION(InvalidOption(arg + " option should follow after --createRandomTest."));
+		};
```
Comment (pirapira) [inline test/tools/libtesteth/Options.cpp]: I believe `--debug` option can appear before `--createRandomTest`.  So the check should not happen during the parsing.  After the parsing, you can detect it.

File: test/tools/fuzzTesting/createRandomTest.cpp (hunk 11 above)
Comment (winsvega) [inline test/tools/fuzzTesting/createRandomTest.cpp]: that would touch too many other unrelated code. like execution of all of the other test suites.

Comment (pirapira) [inline test/tools/fuzzTesting/createRandomTest.cpp]: OK.  This will be kept track in https://github.com/ethereum/cpp-ethereum/issues/4324

File: test/tools/fuzzTesting/fuzzHelper.cpp [hunk 16]
```diff
@@ -303,7 +303,7 @@ std::string RandomCode::generate(int _maxOpNumber, RandomCodeOptions _options)
 	return code;
 }
 
-std::string RandomCode::randomUniIntHex(u256 _maxVal, u256 _minVal)
+std::string RandomCode::randomUniIntHex(u256 _maxVal, u256 const _minVal)
```
Comment (pirapira) [inline test/tools/fuzzTesting/fuzzHelper.cpp]: With `const&`, some memory would be saved.

File: test/tools/libtesteth/Options.cpp [hunk 17]
```diff
@@ -238,12 +232,9 @@ Options::Options(int argc, char** argv)
 				g_logVerbosity = indentLevelInt;
 		}
 		else if (arg == "--createRandomTest")
-		{
 			createRandomTest = true;
-			seenCreateRandomTest = true;
-		}
 		else if (arg == "--debug")
-			throwIfNotRandomTest();
+			randomDebug = true;
```
Comment (pirapira) [inline test/tools/libtesteth/Options.cpp]: By the way, can we replace this `--debug` with `--verbosity 30` or something?

File: test/tools/jsontests/StateTests.cpp [hunk 18]
```diff
@@ -49,7 +49,7 @@ void doStateTests(json_spirit::mValue& _v, bool _fillin)
 		string testname = i.first;
 		json_spirit::mObject& o = i.second.get_obj();
 
-		if (_fillin)
+		if (_fillin && !Options::get().createRandomTest)
```
Comment (pirapira) [inline test/tools/jsontests/StateTests.cpp]: How many combinations are possible here?
* [ ] _fillin && !createRandom
* [ ] _fillin && createRandom
* [ ] !_fillin && !createRandom
* [ ] !_fillin && createRandom

We should try to remove unnecessary combinations or throw errors on meaningless combinations.

Comment (winsvega) [inline test/tools/jsontests/StateTests.cpp]: what? why? 
if it is a filler then it is being checked that testname in json object match the test filename. 
unless it is a filling of a randomstate test that does not have a file.

Comment (pirapira) [inline test/tools/jsontests/StateTests.cpp]: I don't understand.  Which of the fours possibilities are supposed to happen here?

File: test/tools/fuzzTesting/createRandomTest.cpp [hunk 19]
```diff
@@ -42,71 +42,28 @@ int checkRandomTest(std::function<void(json_spirit::mValue&, bool)> _doTests, js
...
 {
-	std::string testSuite;
-	std::string testFillString;
-	json_spirit::mValue testmValue;
-	bool checktest = false;
-	bool filldebug = false;
 	bool debug = false;
-	bool filltest = false;
-
+	TestOutputHelper::initTest(1);
```
Comment (gumb0) [inline test/tools/fuzzTesting/createRandomTest.cpp]: Do you need this? You've made the default to be 1

File: test/tools/libtesteth/TestOutputHelper.cpp (hunk 2 above)
Comment (gumb0) [inline test/tools/libtesteth/TestOutputHelper.cpp]: Where was that?  Maybe you shouldn't use the class `TestOutputHelper` that always works with test sequence for the case when `m_maxTests` doesn't make sense

File: test/tools/libtesteth/TestHelper.cpp [hunk 20]
```diff
@@ -560,7 +561,10 @@ void addClientInfo(json_spirit::mValue& _v, std::string const& _testSource)
 		}
 
 		//prepare the relative src path
-		string source = _testSource.substr(_testSource.rfind("/src/"), _testSource.length());
+		string source = _testSource;
+		string::size_type pos = _testSource.rfind("/src/");
```
Comment (gumb0) [inline test/tools/libtesteth/TestHelper.cpp]: Does this work on Windows?

File: test/tools/libtesteth/Options.h [hunk 21]
```diff
@@ -78,6 +77,12 @@ class Options
 	bool wallet = false;
 	/// @}
 
+	/// CreateRandomTest
+	/// @{
+	bool createRandomTest = false; ///< Generate random test
+	bool randomDebug = false; ///< Debug random test generation
```
Comment (gumb0) [inline test/tools/libtesteth/Options.h]: I don't see this option really used anywhere

File: test/tools/fuzzTesting/createRandomTest.cpp [hunk 22]
```diff
@@ -185,84 +146,91 @@ void fillRandomTest(std::function<void(json_spirit::mValue&, bool)> _doTests, st
...
-			if (types.at(i) == "[0xHASH32]")
-				_test.replace(pos, 10, "0x" + dev::test::RandomCode::rndByteSequence(32));
-			else
-			if (types.at(i) == "[HASH32]")
-				_test.replace(pos, 8, dev::test::RandomCode::rndByteSequence(32));
-			else
-			if (types.at(i) == "[V]")
+			else if (type == "[CODE]")
+			{
+				int random = (int)dev::test::RandomCode::randomUniInt() % 100;
```
Comment (gumb0) [inline test/tools/fuzzTesting/createRandomTest.cpp]: `randomUniInt` has a `_maxVal` parameter which can do this `% 100` for you
(which is a terrible way to generate random integers, not uniform at all)

File: test/tools/fuzzTesting/createRandomTest.cpp [hunk 23]
```diff
@@ -42,71 +42,28 @@ int checkRandomTest(std::function<void(json_spirit::mValue&, bool)> _doTests, js
...
-			if (checktest)
-			{
-				dev::eth::VMFactory::setKind(dev::eth::VMKind::JIT);
-				return checkRandomTest(dev::test::doVMTests, testmValue, debug);
-			}
-			else
-				fillRandomTest(dev::test::doVMTests, (filltest) ? testFillString : c_testExampleVMTest, filldebug);
-		}
-	}
+	if (testSuite == "StateTestsGeneral")
```
Comment (gumb0) [inline test/tools/fuzzTesting/createRandomTest.cpp]: Please don't duplicate these test suite names all over the place, make string constants, probably in a separate file (to be used by other test code)

File: test/tools/fuzzTesting/createRandomTest.cpp [hunk 24]
```diff
@@ -167,10 +124,14 @@ void fillRandomTest(std::function<void(json_spirit::mValue&, bool)> _doTests, st
...
 		_doTests(v, true);
+		for (auto& obj: v.get_obj())
+			if (!obj.second.get_obj().count("post"))
+				BOOST_ERROR("An error occured when executing random state test!");
+		dev::test::addClientInfo(v, "random generated test");
 	}
 	catch(...)
 	{
-		std::cerr << "Test fill exception!";
+		BOOST_ERROR("Error generating random test!");
```
Comment (gumb0) [inline test/tools/fuzzTesting/createRandomTest.cpp]: Which exceptions can happen here? You could display some info from exception object instead of just "unkown error" message

File: test/tools/fuzzTesting/createRandomTest.cpp [hunk 25]
```diff
@@ -185,84 +146,91 @@ void fillRandomTest(std::function<void(json_spirit::mValue&, bool)> _doTests, st
...
-			else
-			if (types.at(i) == "[HASH32]")
-				_test.replace(pos, 8, dev::test::RandomCode::rndByteSequence(32));
-			else
-			if (types.at(i) == "[V]")
+			else if (type == "[CODE]")
+			{
+				int random = (int)dev::test::RandomCode::randomUniInt() % 100;
+				if (random < 90)
+					replace = "0x"+dev::test::RandomCode::generate(5, options);
```
Comment (gumb0) [inline test/tools/fuzzTesting/createRandomTest.cpp]: Spaces around `+`

File: test/tools/fuzzTesting/fuzzHelper.cpp [hunk 26]
```diff
@@ -303,15 +303,15 @@ std::string RandomCode::generate(int _maxOpNumber, RandomCodeOptions _options)
...
 {
 	if (_maxVal == 0)
 		_maxVal = std::numeric_limits<uint64_t>::max();
 	refreshSeed();
 	int rand = randUniIntGen() % 100;
 	if (rand < 50)
-		return toCompactHexPrefixed((u256)randUniIntGen() % _maxVal);
-	return toCompactHexPrefixed((u256)randUInt64Gen() % _maxVal);
+		return toCompactHexPrefixed(_minVal + (u256)randUniIntGen() % _maxVal, 1);
```
Comment (gumb0) [inline test/tools/fuzzTesting/fuzzHelper.cpp]: This is not correct at all, you'll get numbers between `_minVal` and `_minVal + maxVal - 1`

The optimal way I suggest is to create `boostIntDistrib(_minVal, _maxVal)` here and use it

File: test/tools/libtesteth/TestOutputHelper.cpp [hunk 27]
```diff
@@ -56,7 +57,8 @@ void TestOutputHelper::initTest(json_spirit::mValue& _v)
 	NoProof::init();
 	m_timer.restart();
 	m_currentTestCaseName = boost::unit_test::framework::current_test_case().p_name;
-	std::cout << "Test Case \"" + m_currentTestCaseName + "\": " << std::endl;
+	if (!Options::get().createRandomTest)
```
Comment (gumb0) [inline test/tools/libtesteth/TestOutputHelper.cpp]: Why doesn't `TestOutputHelper::initTest(json_spirit::mValue& _v)` just call `initTest(_v.get_obj().size())` ?

File: test/tools/jsontests/StateTests.cpp (hunk 18 above)
Comment (gumb0) [inline test/tools/jsontests/StateTests.cpp]: As I understand all four can happen

(which is terrible, one piece of code has too many responsibilities, it should be 4 different pieces of code/functions)

File: test/tools/fuzzTesting/createRandomTest.cpp [hunk 28]
```diff
@@ -185,84 +146,91 @@ void fillRandomTest(std::function<void(json_spirit::mValue&, bool)> _doTests, st
...
+				int random = (int)dev::test::RandomCode::randomUniInt() % 100;
+				if (random < 90)
+					replace = dev::test::RandomCode::randomUniIntHex(std::numeric_limits<uint32_t>::max());
+				else
+					replace = "0x00";
+			}
+			else if (type == "[GASLIMIT]")
+				replace = dev::test::RandomCode::randomUniIntHex(dev::u256("36028797018963967"), dev::u256("18014398509481983"));
+			else if (type == "[GASPRICE]")
+				replace = dev::test::RandomCode::randomUniIntHex(dev::u256("5"));
```
Comment (gumb0) [inline test/tools/fuzzTesting/createRandomTest.cpp]: More efficient to construct it from integer: `u256(5)`

File: test/tools/libtesteth/TestHelper.cpp (hunk 20 above)
Comment (winsvega) [inline test/tools/libtesteth/TestHelper.cpp]: test filling does not supported on windows

File: test/tools/fuzzTesting/createRandomTest.cpp (hunk 23 above)
Comment (winsvega) [inline test/tools/fuzzTesting/createRandomTest.cpp]: so far this is the only exception.

Comment (pirapira) [inline test/tools/fuzzTesting/createRandomTest.cpp]: @winsvega The suggestion here was `string const g_StateTestsGeneral = "StateTestsGeneral";` somewhere, and reuse the variable all over the place.

Comment (gumb0) [inline test/tools/fuzzTesting/createRandomTest.cpp]: Yeah, the name should be `c_` prefix according to Coding Standards

Comment (winsvega) [inline test/tools/fuzzTesting/createRandomTest.cpp]: done please review other changes so I make all the fixes at once

File: test/tools/fuzzTesting/createRandomTest.cpp (hunk 24 above)
Comment (winsvega) [inline test/tools/fuzzTesting/createRandomTest.cpp]: done

File: test/tools/libtesteth/TestOutputHelper.cpp (hunk 27 above)
Comment (winsvega) [inline test/tools/libtesteth/TestOutputHelper.cpp]: updated. thanks.

File: test/tools/jsontests/StateTests.cpp (hunk 18 above)
Comment (winsvega) [inline test/tools/jsontests/StateTests.cpp]: stop being hilarious. we don't need 4 functions here. 
just one condition is being checked here:
```  if (mode is filler  and we are not in random test  )  then  (check that filename is the same as test name) ```

Comment (winsvega) [inline test/tools/jsontests/StateTests.cpp]: the design of the tests was originally to run a function doTests  and  if passed a flag  filling it will generate the test, otherwise it will run a test. because the code for testing and generating a test is pretty much the same.

Comment (pirapira) [inline test/tools/jsontests/StateTests.cpp]: What about removing the `if` and say
```
if (TestOutputHelper::testFillerFileExists())
    BOOST_REQUIRE_MESSAGE(testname + "Filler.json" == TestOutputHelper::testFillerFileName());
```

File: test/tools/fuzzTesting/createRandomTest.cpp [hunk 29]
```diff
@@ -185,84 +145,102 @@ void fillRandomTest(std::function<void(json_spirit::mValue&, bool)> _doTests, st
...
-				_test.replace(pos, 10, "0x" + dev::test::RandomCode::rndByteSequence(32));
-			else
-			if (types.at(i) == "[HASH32]")
-				_test.replace(pos, 8, dev::test::RandomCode::rndByteSequence(32));
-			else
-			if (types.at(i) == "[V]")
+			else if (type == "[CODE]")
 			{
-				int random = (int)dev::test::RandomCode::randomUniInt() % 100;
+				int random = dev::test::RandomCode::randomPercent();
```
Comment (gumb0) [inline test/tools/fuzzTesting/createRandomTest.cpp]: A lot of `dev::test::RandomCode::`, maybe it's time for `using namespace dev::test::RandomCode;` ?

File: test/tools/fuzzTesting/fuzzHelper.cpp [hunk 30]
```diff
@@ -507,18 +512,58 @@ void RandomCodeOptions::setWeight(eth::Instruction _opCode, int _weight)
...
 }
 
-Address RandomCodeOptions::getRandomAddress() const
+Address RandomCodeOptions::getRandomAddress(AddressType _type) const
 {
-	if (addressList.size() > 0)
+	int index;
+	int random;
+	int probability;
+	switch(_type)
```
Comment (gumb0) [inline test/tools/fuzzTesting/fuzzHelper.cpp]: space after keywords

File: test/tools/fuzzTesting/fuzzHelper.cpp [hunk 31]
```diff
@@ -303,15 +305,13 @@ std::string RandomCode::generate(int _maxOpNumber, RandomCodeOptions _options)
...
+std::string RandomCode::randomUniIntHex(u256 const& _minVal, u256 _maxVal)
 {
 	if (_maxVal == 0)
 		_maxVal = std::numeric_limits<uint64_t>::max();
 	refreshSeed();
-	int rand = randUniIntGen() % 100;
-	if (rand < 50)
-		return toCompactHexPrefixed((u256)randUniIntGen() % _maxVal);
-	return toCompactHexPrefixed((u256)randUInt64Gen() % _maxVal);
+	u256 value = _minVal + (u256)randUInt64Gen() % (_maxVal - _minVal);
```
Comment (gumb0) [inline test/tools/fuzzTesting/fuzzHelper.cpp]: Now it's in `[_minVal; _maxVal-1]`, not what I would expect from function signature like this

File: test/tools/fuzzTesting/fuzzHelper.cpp [hunk 32]
```diff
@@ -507,18 +512,58 @@ void RandomCodeOptions::setWeight(eth::Instruction _opCode, int _weight)
...
+	int random;
+	int probability;
+	switch(_type)
 	{
-		int index = (int)RandomCode::randomUniInt() % addressList.size();
-		return addressList[index];
+		case AddressType::CALLONLY:
+			index = (int)RandomCode::randomUniInt() % callAddressList.size();
+			return callAddressList[index];
+		break;
```
Comment (gumb0) [inline test/tools/fuzzTesting/fuzzHelper.cpp]: `break` should be on the same level of indentation as the code inside `case`
`case` is usually on the same level of indentation as `{` in our code

File: test/tools/fuzzTesting/fuzzHelper.cpp [hunk 33]
```diff
@@ -507,18 +512,58 @@ void RandomCodeOptions::setWeight(eth::Instruction _opCode, int _weight)
...
+	}
 }
 
-Address RandomCodeOptions::getRandomAddress() const
+Address RandomCodeOptions::getRandomAddress(AddressType _type) const
 {
-	if (addressList.size() > 0)
+	int index;
+	int random;
```
Comment (gumb0) [inline test/tools/fuzzTesting/fuzzHelper.cpp]: `random` & `probability` have no reason to be in the outer scope of the function, make them local to the `case` code (by putting code inside `case` into braces `{}`)

Comment (gumb0) [inline test/tools/fuzzTesting/fuzzHelper.cpp]: I'd do the same with `index`, too, actually (not reusing it in different `case`s)

File: test/tools/libtesteth/Options.cpp [hunk 34]
```diff
@@ -279,6 +279,19 @@ Options::Options(int argc, char** argv)
...
 
+	//check restrickted options
+	if (createRandomTest)
+	{
+		string optionlist = "trValueIndex, trGasIndex, trDataIndex, nonetwork, singleTest, performance, quadratic, memory, inputLimits, bigData, wallet, stats, filltests, fillchain";
+		if (trValueIndex >= 0 || trGasIndex >= 0 || trDataIndex >= 0 || nonetwork || singleTest
+			|| 	performance || quadratic || memory || inputLimits || bigData || wallet
+			|| 	stats || filltests || fillchain)
+		{
+			cerr << "--createRandomTest could not be used with one of the options: " + optionlist << endl;
```
Comment (gumb0) [inline test/tools/libtesteth/Options.cpp]: Also I don't see why you need `optionlist` in a variable, just output it here as a literal
